python/cucim/src/cucim/skimage/morphology/convex_hull.py

```diff
@@ -0,0 +1,315 @@
+    # Note: This fix is not yet in upstream scikit-image as of v0.25.
+    original_shape = image.shape
+    image = cp.squeeze(image)
+    if image.ndim < 2:
+        return image
+
+    if image.size < cpu_fallback_threshold:
+        # Fallback to pure CPU implementation
+        from skimage import morphology as morphology_cpu
+
+        convex_image = cp.asarray(
+            morphology_cpu.convex_hull_image(
+                cp.asnumpy(image),
+                offset_coordinates=offset_coordinates,
+                tolerance=tolerance,
+                include_borders=include_borders,
+            )
+        )
+        return convex_image.reshape(original_shape)
+
+    if not scipy_available:
+        raise ImportError(
+            "This function requires SciPy, but it could not import: "
+            "scipy.spatial.ConvexHull, scipy.spatial.QhullError"
+        )
+    if not unique_rows_available:
+        raise ImportError(
+            "This function requires skimage.util.unique_rows, but it could "
+            "not be imported."
+        )
+
+    if not include_borders:
+        raise NotImplementedError(
+            "Only the `include_borders=True` case is implemented"
+        )
+
+    ndim = image.ndim
+    if not omit_empty_coords_check and cp.count_nonzero(image) == 0:
+        warn(
+            "Input image is entirely zero, no valid convex hull. "
+            "Returning empty image",
+            UserWarning,
+        )
+        return np.zeros(image.shape, dtype=bool)
+
+    if image.dtype != cp.dtype(bool):
+        if image.dtype == cp.uint8:
+            # bool are actually already stored as uint8 so can use a view to
+            # avoid a copy
+            image = image.view(bool)
+        else:
+            image = image.astype(bool)
+
+    # xor with eroded version to keep only edge pixels of the binary image
+    image_boundary = cp.bitwise_xor(image, ndi.binary_erosion(image, 3))
+
+    coords = cp.stack(cp.nonzero(image_boundary), axis=-1)
+
+    # Add a vertex for the middle of each pixel edge
+    if offset_coordinates:
+        offsets = _offsets_diamond(image.ndim)
+        coords = (coords[:, np.newaxis, :] + offsets).reshape(-1, ndim)
+
+    coords = cp.asnumpy(coords)
+
+    # repeated coordinates can *sometimes* cause problems in
+    # scipy.spatial.ConvexHull, so we remove them.
+    coords = unique_rows(coords)
+
+    # Find the convex hull
+    try:
+        hull = ConvexHull(coords)
+    except QhullError as err:
+        warn(
+            f"Failed to get convex hull image. "
+            f"Returning empty image, see error message below:\n"
+            f"{err}"
+        )
+        return cp.zeros(image.shape, dtype=bool)
+
+    coord_dtype = cp.min_scalar_type(max(image.shape))
+    if float64_computation:
+        float_dtype = cp.float64
+    else:
+        # float32 will be used if coord_dtype is <= 16-bit
+        # otherwise, use float64
```

Comment:
This comment is slightly misleading as promote_types behavior depends on the specific dtype, not just its size. It would be better to explicitly document the exact promotion rules or use more explicit logic.

Reply:
I can try to clarify the comment

`coord_dtype` is not provided by the user but determined on line 228 above. For integer inputs, `cp.min_scalar_type` returns an appropriately sized unsigned integer type. If that type is uint8 or uint16 then `promote_types` will give float32. If it is uint32 or uint64, it gives float64.

Reply:
comment has now been updated